### PR TITLE
activate by timestamp

### PR DESCRIPTION
**Pull request by andrewsource147**

### CODE REVIEW

File: programs/merkle-distributor/src/instructions/new_distributor.rs [hunk 1]
```diff
@@ -134,18 +136,26 @@ pub fn handle_new_distributor(
     distributor.clawback_receiver = ctx.accounts.clawback_receiver.key();
     distributor.admin = ctx.accounts.admin.key();
     distributor.clawed_back = false;
-    distributor.enable_slot = enable_slot;
```
Comment (codewithgun) [inline programs/merkle-distributor/src/instructions/new_distributor.rs:137]: Missing`distributor.activation_type = activation_type`

Comment (codewithgun) [inline programs/merkle-distributor/src/instructions/new_distributor.rs:137]: Missing `distributor.activation_type = activation_type`

File: programs/merkle-distributor/src/state/merkle_distributor.rs [hunk 2]
```diff
@@ -58,12 +71,66 @@ pub struct AirdropBonus {
...
+    pub vesting_duration: u64,
     /// total bonus
     pub total_claimed_bonus: u64,
 }
 
+pub struct ActivationHandler {
+    /// current slot or current timestamp
+    pub curr_time: u64,
+    /// activation slot or activation timestamp
+    pub activation_time: u64,
```
Comment (codewithgun) [inline programs/merkle-distributor/src/state/merkle_distributor.rs:83]: Maybe we update all `activation_slot` / `activation_time` interchangeable variable name to `activation_period` which the name itself is not bounded to time/slot?

File: programs/merkle-distributor/src/state/merkle_distributor.rs [hunk 3]
```diff
@@ -58,12 +71,66 @@ pub struct AirdropBonus {
...
+    pub fn get_bonus_for_a_claimaint(&self, max_bonus: u64) -> Result<u64> {
+        let curr_time = self.curr_time;
+        let start_time = self.activation_time;
+        let end_time = self.airdrop_bonus.vesting_duration.safe_add(start_time)?;
+
+        if curr_time >= start_time {
+            if curr_time >= end_time {
+                Ok(max_bonus)
+            } else {
+                let slot_into_unlock = curr_time.safe_sub(start_time)?;
```
Comment (codewithgun) [inline programs/merkle-distributor/src/state/merkle_distributor.rs:105]: What about `duration_into_unlock` as the variable name?

File: programs/merkle-distributor/src/state/merkle_distributor.rs [hunk 4]
```diff
@@ -58,12 +71,66 @@ pub struct AirdropBonus {
...
+                let total_unlock_slot = self.airdrop_bonus.vesting_duration;
```
Comment (codewithgun) [inline programs/merkle-distributor/src/state/merkle_distributor.rs:106]: What about `total_unlock_duration` as the variable name?

File: programs/merkle-distributor/src/instructions/new_distributor.rs (hunk 1 above)
Comment (andrewsource147) [reply programs/merkle-distributor/src/instructions/new_distributor.rs:137]: fixed

File: programs/merkle-distributor/src/state/merkle_distributor.rs (hunk 3 above)
Comment (andrewsource147) [reply programs/merkle-distributor/src/state/merkle_distributor.rs:105]: fixed

File: programs/merkle-distributor/src/state/merkle_distributor.rs (hunk 4 above)
Comment (andrewsource147) [reply programs/merkle-distributor/src/state/merkle_distributor.rs:106]: fixed

File: programs/merkle-distributor/src/state/merkle_distributor.rs (hunk 2 above)
Comment (andrewsource147) [reply programs/merkle-distributor/src/state/merkle_distributor.rs:83]: nice idea, pushed the fix for that

**Review by codewithgun — APPROVED** [review]

**Pull request merged**